internal/db/db.go

```diff
@@ -2,281 +2,452 @@
+		ID: id,
+	}
+	stmt, err := sqlair.Prepare(fmt.Sprintf(getCSRsStmt, db.certificateTable), CertificateRequest{})
+	if err != nil {
+		return nil, err
+	}
+	err = db.conn.Query(context.Background(), stmt, csr).Get(&csr)
+	if err != nil {
+		return nil, err
+	}
+	return &csr, nil
 }
-type User struct {
-	ID          int
-	Username    string
-	Password    string
-	Permissions int
+
+// RetrieveCSRbyCSR gets a given CSR row from the repository using the CSR text.
+func (db *Database) RetrieveCSRbyCSR(csr string) (*CertificateRequest, error) {
+	row := CertificateRequest{
+		CSR: csr,
+	}
+	stmt, err := sqlair.Prepare(fmt.Sprintf(getCSRsStmt, db.certificateTable), CertificateRequest{})
+	if err != nil {
+		return nil, err
+	}
+	err = db.conn.Query(context.Background(), stmt, row).Get(&row)
+	if err != nil {
+		return nil, err
+	}
+	return &row, nil
 }
 
-var ErrIdNotFound = errors.New("id not found")
+// CreateCSR creates a new CSR entry in the repository. The string must be a valid CSR and unique.
+func (db *Database) CreateCSR(csr string) error {
+	if err := ValidateCertificateRequest(csr); err != nil {
+		return errors.New("csr validation failed: " + err.Error())
+	}
+	stmt, err := sqlair.Prepare(fmt.Sprintf(createCSRStmt, db.certificateTable), CertificateRequest{})
+	if err != nil {
+		return err
+	}
+	row := CertificateRequest{
+		CSR: csr,
+	}
+	err = db.conn.Query(context.Background(), stmt, row).Run()
+	if err != nil {
+		return err
+	}
+	return nil
+}
 
-// RetrieveAllCSRs gets every CertificateRequest entry in the table.
-func (db *Database) RetrieveAllCSRs() ([]CertificateRequest, error) {
-	rows, err := db.conn.Query(fmt.Sprintf(queryGetAllCSRs, db.certificateTable))
+// AddCertificateChainToCSRbyCSR adds a new certificate chain to a row for a given CSR string.
+func (db *Database) AddCertificateChainToCSRbyCSR(csr string, cert string) error {
+	err := ValidateCertificate(cert)
 	if err != nil {
-		return nil, err
+		return errors.New("cert validation failed: " + err.Error())
+	}
+	err = CertificateMatchesCSR(cert, csr)
+	if err != nil {
+		return errors.New("cert validation failed: " + err.Error())
+	}
+	certBundle := sanitizeCertificateBundle(cert)
+	stmt, err := sqlair.Prepare(fmt.Sprintf(updateCSRStmt, db.certificateTable), CertificateRequest{})
+	if err != nil {
+		return err
+	}
+	newRow := CertificateRequest{
+		CSR:              csr,
+		CertificateChain: certBundle,
+		RequestStatus:    "Active",
+	}
+	err = db.conn.Query(context.Background(), stmt, newRow).Run()
+	if err != nil {
+		return err
 	}
+	return nil
+}
 
-	var allCsrs []CertificateRequest
-	defer rows.Close()
-	for rows.Next() {
-		var csr CertificateRequest
-		if err := rows.Scan(&csr.ID, &csr.CSR, &csr.Certificate); err != nil {
-			return nil, err
-		}
-		allCsrs = append(allCsrs, csr)
+// AddCertificateChainToCSRbyID adds a new certificate chain to a row for a given row ID.
+func (db *Database) AddCertificateToCSRbyID(id int, cert string) error {
```

Comment:
Do we use this anywhere?

Reply:
yes, in the certificate request handler tests